[WIP] To see CI errors

T-Gro opened the pull request.

edgarfgp commented:
@T-Gro  Would it make sense to add some tests for lowercase DUs https://github.com/fsharp/fslang-design/blob/main/FSharp-7.0/FS-1126-allow-lower-case-du-cases%20when-require-qualified-access-is%20specified.md

```fsharp
[<RequireQualifiedAccess>]
type DU =
     | an of int
     | B of string
     | C
     | ``D`` of bool
     | ``d``
```

T-Gro commented:
Still a draft, I have a regression with losing 'typecheck error FS0912: This declaration element is not permitted in an augmentation '
(i.e. the error should be there, but change in this PR drops it and a different one appears later)

vzarytovskii commented on src/Compiler/AbstractIL/il.fs:
```diff
@@ -1451,6 +1463,9 @@ type ILLocalDebugInfo =
     }
 
+    override x.ToString() =
+        (fst x.Range).ToString() + "-" + (snd x.Range).ToString()
```
Each range (type) access may involve filesystem access, I would be careful here.

vzarytovskii reviewed (commented):
Also, probably worth including debug display for all types you've added tostrings to?

vzarytovskii commented on src/Compiler/AbstractIL/il.fs:
```diff
@@ -1985,7 +2018,7 @@ type ILMethodDef
     member x.IsZeroInit = x.MethodBody.IsZeroInit
 
-    member md.CallingSignature =
+    member md.GetCallingSignature() =
```
Wouldn't make it lazy make more sense? In case we need to access it multiple times, we don't compute it multiple times?

T-Gro replied:
It's just a record-creating no-op, and only used when a match by name is already found.
Better to keep the cost just in time, rather then allocate additional lazy<> each time (even when not used in the end)

KevinRansom reviewed (approved):
Looks good.

T-Gro commented on src/Compiler/AbstractIL/il.fs:
```diff
@@ -422,6 +424,8 @@ type AssemblyRefData =
         assemRefLocale: Locale option
     }
 
+    override x.ToString() = x.assemRefName
```
I think that this (il.fs) file decided to standardize on "member x." :

![image](https://github.com/dotnet/fsharp/assets/46543583/91862013-f334-4980-b04d-0f8a1a5f5d3f)

T-Gro commented:
> Also, probably worth including debug display for all types you've added tostrings to?

Do we need to?

"Thus, if you have overridden ToString(), you do not need to use [DebuggerDisplayAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.debuggerdisplayattribute)"

vzarytovskii commented on src/Compiler/AbstractIL/ilx.fs:
```diff
@@ -58,10 +72,15 @@ type IlxUnionSpec =
+
     member x.FieldDef idx fidx = x.Alternative(idx).FieldDef(fidx)
 
+    override x.ToString() = x.TypeRef.Name
```
Can `TypeRef` here be null for some reason?

T-Gro replied:
Here it cannot, it is always filled.

T-Gro commented:
(copying also here)

I have now covered the multi-file scenario as well:

https://github.com/dotnet/fsharp/pull/15343/files#diff-612f0ce4418f76f1f79c4bd03811dde6e6424025ecb489c4a3d6a104ebeb7303R316


It does work as expected:
- neither type's nor module's init is called if nothing from the file is referenced by main program
- non-main content is initialized on first access to anything in that file (following the existing behavior on what constitutes an access or doesn't)
     - The file's static init contents are accessed in order:
         1. All module-level bindings/do's which happened before the type definition
         2. Then the actual static content of the type
         3. Then module-level content which was declared below the type definition

T-Gro commented on src/Compiler/AbstractIL/il.fs:
```diff
@@ -1451,6 +1463,10 @@ type ILLocalDebugInfo =
     }
 
+    override x.ToString() =
+        let firstLabel, secondLabel = x.Range
```
It is not the "range" type, here it's just a pair of ints.
(I have also added very explicit %i below in the sprintf call to ensure those remain to be plain integers)

vzarytovskii commented on src/Compiler/Checking/CheckDeclarations.fs:
```diff
@@ -744,23 +746,29 @@ module MutRecBindingChecking =
     type TyconBindingPhase2A =
       /// An entry corresponding to the definition of the implicit constructor for a class
-      | Phase2AIncrClassCtor of IncrClassCtorLhs
+      | Phase2AIncrClassCtor of StaticCtorInfo * IncrClassCtorInfo option
```
Why does it now contain static constructor info as well now?
Either comment should be fixed, or new union constructor introduced.

T-Gro replied:
Before this PR, info about static cctor used to be part of IncrClassCtorInfo, since those were the only types where static ctor actions were allowed.
This PR splits them into two.

=> code is correct, comment is wrong, fixing comment right now.

vzarytovskii commented on src/Compiler/CodeGen/IlxGen.fs:
```diff
@@ -1859,29 +1859,29 @@ let MergePropertyDefs m ilPropertyDefs =
 type TypeDefBuilder(tdef: ILTypeDef, tdefDiscards) =
-    let gmethods = ResizeArray<ILMethodDef>(0)
-    let gfields = ResizeArray<ILFieldDef>(0)
+    let gmethods = ResizeArray<ILMethodDef>(tdef.Methods.AsList())
```
I guess the idea before was that we don't grow the list with methods from typedef, and only make those in `Close`. I presume it's still *should* be fine, but i'd double check on big types.

vzarytovskii reviewed (commented):
We need to (manually) test the following scenario:
1. Build the library with compiler from this branch
2. Load it in project with older compiler, which does not have this feature.
3. See that we generate IL + pickled info, which is loadable and accessible by earlier compilers.

T-Gro commented on src/Compiler/TypedTree/TypedTreePickle.fs:
```diff
@@ -1790,32 +1790,81 @@ let u_istype st =
-    | TFSharpObjectRepr r  -> p_byte 1 st; p_byte 3 st; p_tycon_objmodel_data r st; false
-    | TMeasureableRepr ty  -> p_byte 1 st; p_byte 4 st; p_ty ty st; false
-    | TNoRepr              -> p_byte 0 st; false
+    // Records
```
Union cases with static fields follow a new branch and are added to the format.

vzarytovskii replied:
Won't this then make old compilers fail on loading the pickled data?

auduchinok replied:
I'd expect the older compilers (and IDE versions) to fail [here](https://github.com/dotnet/fsharp/pull/15343/files#diff-ba2d0f10532876848717bf5c70624a46101cf86194da89d711bb57e99bdcafceR2101) and [here](https://github.com/dotnet/fsharp/pull/15343/files#diff-ba2d0f10532876848717bf5c70624a46101cf86194da89d711bb57e99bdcafceR2304).

vzarytovskii replied:
> I'd expect the older compilers (and IDE versions) to fail [here](https://github.com/dotnet/fsharp/pull/15343/files#diff-ba2d0f10532876848717bf5c70624a46101cf86194da89d711bb57e99bdcafceR2101) and [here](https://github.com/dotnet/fsharp/pull/15343/files#diff-ba2d0f10532876848717bf5c70624a46101cf86194da89d711bb57e99bdcafceR2304).

Yep, exactly what I though, thanks.

T-Gro replied:
@auduchinok  is correct, it leads to:

Problem reading assembly 'D:\StaticLetTesting\UnionShowCase.exe': The exception has been reported. This internal exception should now be caught at an error recovery point on the stack. Original message: Error opening binary file 'D:\StaticLetTesting\UnionShowCase.exe': Error reading/writing metadata for the F# compiled DLL 'D:\StaticLetTesting\UnionShowCase.exe'. Was the DLL compiled with an earlier version of the F# compiler? (**error: 'u_tycon_repr'**).)	scratchproject	D:\scratchproject\FSC	1	Active

T-Gro replied:
==> I will mark this PR against merging, and check how the format could be kept.

T-Gro commented:
I added a change that will make Fsharp.Core compilation fail if a feature requiring new pickling codes is used. We have discussed this with @dsyme and @vzarytovskii and Fsharp.Core is the only library where we want to support compilation on older tooling - it will fail if new constructs are used in it. Otherwise, it should not be a reason to stop language evolution as library authors can suggest to update tooling.

Methodology:
I have added a Discriminated Union to SI.fs in Fsharp.Core that was using "static let" fields and actions, marked under
```
#if !PROTO
```
I have updated Fsharp.Core to use preview version of the language.

I ran the `build.cmd` script.

The build has failed with the following two errors when pickling FSharp.Core:
```
D:\fsharp\src\FSharp.Core\SI.fs(236,16): error FS3569: Newly added pickle state cannot be used in FSharp.Core, since it must be working in older compilers+tooling as well. The time window is at least 3 years after feature introduction. Violation: fields in union . Context:  sizeOfTCachedperTyparInstMutableCounter [D:\fsharp\src\FSharp.
Core\FSharp.Core.fsproj::TargetFramework=netstandard2.0]
D:\fsharp\src\FSharp.Core\SI.fs(236,16): error FS3569: Newly added pickle state cannot be used in FSharp.Core, since it must be working in older compilers+tooling as well. The time window is at least 3 years after feature introduction. Violation: fields in union . Context:  sizeOfTCachedperTyparInstMutableCounter [D:\fsharp\src\FSharp.
Core\FSharp.Core.fsproj::TargetFramework=netstandard2.1]
```


(once for NS2.0, and once for NS2.1)


I am now keeping the check itself in the code, and rolling back the changes for making Fsharp.Core fail for demonstrative reasons.

T-Gro commented:
> What about a test in EmittedIL to verify what initialization takes place under the hood?

Will add, one single file and one multi file (compilation strategy differs)

T-Gro commented:
This is ready now

edgarfgp commented:
@T-Gro I'm curious to see if these changes will help with https://github.com/dotnet/fsharp/issues/11929 ?

T-Gro commented:
> @T-Gro I'm curious to see if these changes will help with #11929 ?

I am afraid this is orthogonal @edgarfgp , the issue still remains.
